### PR TITLE
[SPARK-34868][SQL] Support divide an year-month interval by a numeric

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
1. Add new expression `DivideYMInterval` which multiplies a `YearMonthIntervalType` expression by a `NumericType` expression including ByteType, ShortType, IntegerType, LongType, FloatType, DoubleType, DecimalType.
2. Extend binary arithmetic rules to support `year-month interval / numeric`.

### Why are the changes needed?
To conform the ANSI SQL standard which requires such operation over year-month intervals:
Screenshot 2021-03-25 at 18 44 58

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
By running new tests:
```
$ build/sbt "test:testOnly *IntervalExpressionsSuite"
$ build/sbt "test:testOnly *ColumnExpressionSuite"
```

### CODE REVIEW

**Comment by tanelk** [comment]
Perhaps a bit broader question. When adding new expressiones, should we keep in mind places like `PushFoldableIntoBranches#supportedBinaryExpression`? Or should they be handled separately?

**Comment by MaxGekk** [comment]
> Or should they be handled separately?

@tanelk Separately. There is the JIRA https://issues.apache.org/jira/browse/SPARK-34841 for that. Need to add new expressions + tests.

**Comment by MaxGekk** [comment]
@yaooqinn @cloud-fan Could you review this PR, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala [hunk 1]
```diff
@@ -341,3 +341,53 @@ case class MultiplyDTInterval(
...
+  override def left: Expression = interval
+  override def right: Expression = num
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(YearMonthIntervalType, NumericType)
+  override def dataType: DataType = YearMonthIntervalType
+
+  @transient
+  private lazy val evalFunc: (Int, Any) => Any = right.dataType match {
+    case LongType => (months: Int, num) =>
+      LongMath.divide(months, num.asInstanceOf[Long], RoundingMode.HALF_UP).toInt
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala:359]: `months` is int, the result of `/` must fit to int too. So, `toInt` should be safe.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala:359]: It's better to put it as a code comment. We can fix it in the next PR.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by yaooqinn** [comment]
late lgtm.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala:359]: I added comments in https://github.com/apache/spark/pull/31972

**Pull request closed without merging**